### ISSUE TITLE
Avoid synthetic ORDER BY clause in window functions if not strictly necessary

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#11021]
jOOQ 3.12 introduced synthetic `ORDER BY` clauses for window functions where they should be required, i.e. where ordering is relevant. In some dialects (e.g. Db2: #8593), the `ORDER BY` clauses are mandatory in these cases. In other dialects (e.g. H2: #8414, Oracle), however, they are not.

A customer has reported that their SQL is no longer working, because they depend on (undocumented?) behaviour where `FIRST_VALUE (x) OVER ()` produces the first value according to the `ORDER BY` clause of:

- The `SELECT` statement
- Another window function

While it may be brittle to rely on this implicit ordering, there's no reason for jOOQ to generate a dummy `ORDER BY` clause, which is equally arbitrary, but produces a different behaviour from that which may have been intended.

This issue will list dialects and window functions where #8593 is rolled back again.

### Mandatory `ORDER BY` clauses per dialect:

| Function | Db2 | HANA | H2 | Oracle | SQL Server | Vertica |
| -- | -- | -- | -- | -- | -- | -- |
| `first_value` | | x | | | x | x |
| `last_value` | | x | | | x | x |
| `nth_value` | | x | | | x | x |
| `lead` | x | x | x | x | x | x |
| `lag` | x | x | x | x | x | x |
| `ntile` | x | x | x | x | x | x |
| `rank` | x | x | x | x | x | x |
| `dense_rank` | x | x | x | x | x | x |
| `percent_rank` | x | x | | x | x | x |
| `cume_dist` | x | x | | x | x | x |
| `row_number` | | | | x | x | |

Not all dialects are listed here. Some dialects (e.g. MySQL, PostgreSQL, SQLite) *never* require `ORDER BY`)

**Comment by lukaseder** [comment]
@katzyn: Is it true, does H2 not require `ORDER BY` for `PERCENT_RANK` and `CUME_DIST`? What's the rationale?

**Comment by katzyn** [comment]
Requirements in H2 are exactly the same as in the SQL Standard:
```
7) If <ntilea>, <lead or lag function>, <rank function type> or ROW_NUMBER
   is specified, then:
  a) If <ntile function>, <lead or lag function>, RANK or DENSE_RANK
     is specified, then the window ordering clause WOC of WDX
     shall be present.
  b) The window framing clause of WDX shall not be present.
```

From my point of view there are no good reasons to allow missing window ordering clause for these two functions to, because `PERCENT_RANK` is always 0 and `CUME_DIST` is always 1 without it, but they are allowed by the Standard.

Current H2 optimizes `RANK` and `DENSE_RANK` to `CAST(1 AS BIGINT)`, `PERCENT_RANK` to `CAST(0 AS DOUBLE PRECISION)` and `CUME_DIST` to `CAST(1 AS DOUBLE PRECISION)` when they are used with constant window ordering clause or without any window ordering clause. Hypothetically jOOQ can do the same when one of these functions without window ordering clause was requested and target DBMS doesn't support it.

**Comment by lukaseder** [comment]
Oh interesting, thanks for the pointer to the standard. I should have checked myself.

> Hypothetically jOOQ can do the same when one of these functions without window ordering clause was requested and target DBMS doesn't support it.

jOOQ could, but that may not have been the intention of the query author. Every time such an optimisation is implemented (in jOOQ), someone will run into the edge case where the actual expression (the window function) would have been more desirable 😁

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.4 (#11029)

**Comment by katzyn** [comment]
Situation with window framing clause is even more obscure, some DBMS accept in when it isn't allowed by the Standard, that's OK, but it is actually used only for some window functions and is silently ignored for others.

**Comment by lukaseder** [comment]
> Situation with window framing clause is even more obscure, some DBMS accept in when it isn't allowed by the Standard, that's OK, but it is actually used only for some window functions and is silently ignored for others.

Well, it could be worse :)

**Comment by lukaseder** [comment]
The informix fix will ship with 3.14.5
